README.md
koa-json-logger
===============

KoaJS HTTP Request/Response/Error JSON format logger.

[![Build Status](https://travis-ci.org/rudijs/koa-json-logger.svg?branch=master)](https://travis-ci.org/rudijs/koa-json-logger)
[![Coverage Status](https://coveralls.io/repos/rudijs/koa-json-logger/badge.png?branch=master)](https://coveralls.io/r/rudijs/koa-json-logger?branch=master)
[![NPM version](https://badge.fury.io/js/koa-json-logger.svg)](http://badge.fury.io/js/koa-json-logger)
[![Dependency Status](https://gemnasium.com/rudijs/koa-json-logger.svg)](https://gemnasium.com/rudijs/koa-json-logger)

The actual logging is done with [node-bunyan](https://github.com/trentm/node-bunyan)

Credits and inspired by:

* [KoaJS](https://github.com/koajs)
* [express-winston](https://github.com/heapsource/express-winston)
* [node-bunyan](https://github.com/trentm/node-bunyan)
* [koa-bunyan](https://github.com/ivpusic/koa-bunyan)
* [KoaJS Logger](https://github.com/koajs/logger)


Code review, suggestions and pull requests very much welcome - thanks!

## Overview

The basic goal of this module is to create log entries in JSON format for each HTTP request.

Success (<400) log entries will contain Request and Response details.

Error (>=400) log entries will contain Request, Response and Error details.

Success and Error logs will have their own file (two files).

In NODE_ENV development errors will also be output to the console.

500 error responses are logged in full detail but the HTTP user response will always be only 'Internal Server Error'.

Less than 500 and greater than 500 error responses are logged and passed through as is to the user in the HTTP response.

Higher level goals are:

- Create uniform success and error log entries in JSON format for centralized logging
- In particular tested and used with the ELK stack (Logstash, Elasticsearch and Kibana)
- A unique ID (RFC4122 uuid v4) is also created for each log entry.
- This uid can be used in other application logs, which end up in ELK, so you can correlate the request or error with other application log entries.
- Support [JSON API](http://jsonapi.org/) response type

Below will be some sample log entries in pretty print and also some screen shots of how they look in Kibana.

## Install

`npm install koa-json-logger`

## Usage

`var koaJsonLogger = require('koa-json-logger');`

`app.use(koaJsonLogger());`

I suggest it's best to use this middleware high in the middleware stack so any and all downstream errors are logged.

Logs will go into a `log/` directory relative to file that instruments the Koa app, so you'll need to create this folder.

Default use will create two log files:

`log/app.log` will contain req/res log entries.

`log/app_error.log` will contain error log entries.

Log files are not auto rotated, recommend using system file rotation facility such as `logrotate` on Linux or `logadm` on SmartOS/Illumos.

Current config options are:

- `name` (String default: 'app') configures the log file name and name property of the log entry.

- `path` (String default: 'log') configures the log directory (relative) to use.

- `jsonapi` (Boolean default: false) will set the response Content-type to application/vnd.api+json and ensure 500 responses are in JSON API format

Example:

      app.use(koaJsonLogger({
        name: 'myCoolApp',
        path: 'logs',
        jsonapi: true
      }));

When you throw an application error it's best to always use [throw](https://github.com/koajs/koa/blob/master/docs/api/context.md#ctxthrowmsg-status-properties)

Example:

    `this.throw(403, 'Access Denied');`

See the docs for [throw](https://github.com/koajs/koa/blob/master/docs/api/context.md#ctxthrowmsg-status-properties) details.

You can return errors this way (no throw) but without throwing this will end up in the standard log file - *not* the error log file.

So best *not* to do it this way (the error message in the body will *not* be logged):

    // Not a good way to an return error
    this.status = 401;
    this.body = 'Access Denied';

If using [JSON API](http://jsonapi.org/) set `jsonapi: true` - response Content-type header will be set to application/vnd.api+json

You can then throw errors like this for example:

    this.throw(401, {
      message: {
        status: 401,
        title: 'Unauthorized',
        detail: 'No Authorization header found'
      }
    }

The unique ID (uid) set for each request and can be accessed/used with `this.uuid`

For example using Bunyan elsewhere in your application for logging you could use the uid like so:

`logger.info({uid: this.uuid}, 'Application log message here');`

Then you can correlate http request/response/error log entries with you application log entries based on the uid.

Please review the test suite for further details.

## Tests with code coverage report in `test/coverage`

Note: Requires nodes at least v0.11.13 (earlier v0.11 versions may work, have not checked for this).

git clone the full repo: `git clone git@github.com:rudijs/koa-json-logger.git`

`cd koa-json-logger`

`npm install`

`npm test`


## Code Linting

`npm run lint`


## Sample Success Log Entry (pretty print)

	{
	   "name":"rsm-api",
	   "hostname":"dev",
	   "pid":30213,
	   "level":30,
	   "uid":"37154307-aed2-4e10-b702-97ed9e7d4a3b",
	   "req":{
	      "url":"/users/539d77e8cd4e834b710a103a",
	      "headers":{
		 "host":"127.0.0.1:3000",
		 "connection":"keep-alive",
		 "cache-control":"no-cache",
		 "user-agent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/33.0.1750.152 Safari/537.36",
		 "content-type":"application/vnd.api+json",
		 "accept":"*/*",
		 "accept-encoding":"gzip,deflate,sdch",
		 "accept-language":"en-US,en;q=0.8"
	      },
	      "method":"GET",
	      "ip":"::ffff:127.0.0.1",
	      "protocol":"http",
	      "originalUrl":"/users/539d77e8cd4e834b710a103a",
	      "query":{}
	   },
	   "res":{
	      "statusCode":200,
	      "responseTime":15,
	      "headers":{
		 "cache-control":"no-store, no-cache",
		 "x-content-type-options":"nosniff",
		 "x-download-options":"noopen",
		 "x-xss-protection":"1; mode=block",
		 "x-frame-options":"DENY",
		 "vary":"Accept-Encoding",
		 "content-type":"application/json; charset=utf-8",
		 "content-length":"66"
	      }
	   },
	   "msg":"",
	   "time":"2014-08-21T16:24:24.118Z",
	   "v":0
	}

## Sample Error Log Entry (pretty print)

	{
	   "name":"rsm-api",
	   "hostname":"dev",
	   "pid":30213,
	   "level":50,
	   "uid":"ec874d07-dd94-4298-a4c3-8181a6803d4a",
	   "req":{
	      "url":"/users/539d77e8cd4e834b710a103a",
	      "headers":{
		 "host":"127.0.0.1:3000",
		 "connection":"keep-alive",
		 "cache-control":"no-cache",
		 "user-agent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/33.0.1750.152 Safari/537.36",
		 "content-type":"application/vnd.api+json",
		 "accept":"*/*",
		 "accept-encoding":"gzip,deflate,sdch",
		 "accept-language":"en-US,en;q=0.8"
	      },
	      "method":"GET",
	      "ip":"::ffff:127.0.0.1",
	      "protocol":"http",
	      "originalUrl":"/users/539d77e8cd4e834b710a103a",
	      "query":{}
	   },
	   "res":{
	      "statusCode":401,
	      "responseTime":3,
	      "headers":{
		 "cache-control":"no-store, no-cache",
		 "x-content-type-options":"nosniff",
		 "x-download-options":"noopen",
		 "x-xss-protection":"1; mode=block",
		 "x-frame-options":"DENY",
		 "vary":"Accept-Encoding"
	      }
	   },
	   "err":{
	      "message":{
		 "status":401,
		 "title":"Unauthorized",
		 "detail":"Please sign in to complete this request."
	      },
	      "name":"Error",
	      "stack":"Error: Unauthorized\n    at Object.module.exports [as throw] (/media/crypt2/projects/ride-share-market-api/node_modules/koa/lib/context.js:84:48)\n    at Object.authorization (/media/crypt2/projects/ride-share-market-api/app/middlewares/authorization.js:49:19)\n    at GeneratorFunctionPrototype.next (native)\n    at Object.<anonymous> (/media/crypt2/projects/ride-share-market-api/node_modules/koa-router/node_modules/koa-compose/index.js:29:12)\n    at GeneratorFunctionPrototype.next (native)\n    at next (/media/crypt2/projects/ride-share-market-api/node_modules/koa/node_modules/co/index.js:74:21)\n    at Object.<anonymous> (/media/crypt2/projects/ride-share-market-api/node_modules/koa/node_modules/co/index.js:45:5)\n    at next (/media/crypt2/projects/ride-share-market-api/node_modules/koa/node_modules/co/index.js:90:21)\n    at Object.<anonymous> (/media/crypt2/projects/ride-share-market-api/node_modules/koa/node_modules/co/index.js:45:5)\n    at next (/media/crypt2/projects/ride-share-market-api/node_modules/koa/node_modules/co/index.js:90:21)"
	   },
	   "msg":"",
	   "time":"2014-08-21T16:26:52.338Z",
	   "v":0
	}

## Kibana Success Screenshot example:


![Kibana Success Screenshot](examples/kibana_success_log_entry.png)


## Kibana Error Screenshot example:


![Kibana Error Screenshot](examples/kibana_error_log_entry.png)



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## 1.0.2 - 2016-01-22

### Added
- Package dependency updates.
- Add NPM lint script.

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- Nothing.

## 1.0.1 - 2015-10-20

### Added
- Package dependency updates.

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- Nothing.

## 0.0.22 - 2015-07-28

### Added
- Package dev dependency updates.

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- Nothing.

## 0.0.19 - 2015-03-03

### Added
- Package dependency updates.
- Unit test updates.

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- Nothing.

## 0.0.18 - 2014-11-06

### Added
- Package dependency updates

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- Nothing.

## 0.0.17 - 2014-10-26

### Added
- Package dependency updates

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- Nothing.

## 0.0.15 - 2014-09-13

### Added
- KoaJS dev dependency updated to 0.11.0

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- Nothing.

## 0.0.14 - 2014-08-23

### Added
- Readme updates
- Better support for JSON API format responses

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- Nothing.

## 0.0.13 - 2014-08-22

### Added
- Update to istanbul-harmony npm package
- Readme updates

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- Nothing.

## 0.0.12 - 2014-08-22

### Added
- Readme updates

### Deprecated
- Nothing.

### Removed
- Auto log rotation. Recommended to use system file rotation facility such as logrotate on Linux or logadm on SmartOS/Illumos

### Fixed
- Nothing.

## 0.0.11 - 2014-08-22

### Added
- Readme updates

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- Nothing.

## 0.0.10 - 2014-08-22

### Added
- Readme updates

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- Nothing.

## 0.0.9 - 2014-08-22

### Added
- Ensure default 500 status code for errors
- Readme updates

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- Nothing.

## 0.0.8 - 2014-08-22

### Added
- Greater Request property logging
- Greater Response property logging
- Unique ID for each request add to each log entry and is available application wide for other logs to use.
- Improved Readme details
- Configurable log directory option
- Dependency updates
- JSON API application/vnd.api+json format response support added
- Readme updates

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- Nothing.


## 0.0.7 - 2014-08-10

### Added
- CHANGELOG.md file.

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- Nothing.


test/koa-json-logger.spec.js
'use strict';

var should = require('chai').should(),
  request = require('supertest'),
  sinon = require('sinon'),
  koa = require('koa'),
  fs = require('fs');

var koaJsonLogger = require('../lib/koa-json-logger'),
  testLogFiles = [
    'log/app.log',
    'log/app_error.log'
  ],
  app;

describe('JSON Logger middleware', function () {

  beforeEach(function (done) {
    // clear out test log files
    testLogFiles.forEach(function (logFile) {
      fs.exists(logFile, function (file) {
        if (file) {
          fs.unlinkSync(logFile);
        }
      });
    });
    done();
  });

  beforeEach(function (done) {
    app = koa();
    done();
  });

  describe('success', function () {

    it('should log request and response', function (done) {

      app.use(koaJsonLogger());

      // default route for test
      app.use(function *route1(next) {
        this.body = 'Test Response is OK.';
        yield next;
      });

      request(app.listen())
        .get('/')
        .expect(200)
        .end(function (err, res) {
          if (err) {
            should.not.exist(err);
            return done(err);
          }

          // test http response
          res.text.should.equal('Test Response is OK.');

          // read in log file entry
          fs.readFile('log/app.log', function (err, data) {
            if (err) {
              throw err;
            }

            // test JSON parsed log entry
            var logEntry = JSON.parse(data.toString());

            // bunyan property logging
            logEntry.name.should.equal('app');
            should.exist(logEntry.uid);
            logEntry.uid.should.match(/^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i);

            // request logging
            logEntry.req.method.should.equal('GET');
            logEntry.req.url.should.equal('/');
            should.exist(logEntry.req.headers);

            // response logging
            logEntry.res.statusCode.should.equal(200);
            should.exist(logEntry.res.responseTime);
            logEntry.res.headers['content-type'].should.equal('text/plain; charset=utf-8');

            done();
          });

        });

    });

  });

  describe('error', function () {

    it('should log request, response and error', function (done) {

      app.use(koaJsonLogger());

      // 1st default test route that will catch uncaught downstream errors
      app.use(function *route1(next) {
        yield next;
      });

      // 2nd route. Downstream route/middleware that throws an error
      app.use(function *route2(next) {
        this.throw('Oops! Something blew up.');
        yield next;
      });

      request(app.listen())
        .get('/')
        .expect(500)
        .end(function (err, res) {
          if (err) {
            should.not.exist(err);
            return done(err);
          }

          // should not leak out internal server error messages on 500
          // standard error response for the user
          res.text.should.equal('Internal Server Error');

          // read in log file entry
          fs.readFile('log/app_error.log', function (err, data) {
            if (err) {
              throw err;
            }

            // test JSON parsed log entry
            var logEntry = JSON.parse(data.toString());

            // bunyan property logging
            logEntry.name.should.equal('app');
            should.exist(logEntry.uid);
            logEntry.uid.should.match(/^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i);

            // request logging
            logEntry.req.method.should.equal('GET');
            logEntry.req.url.should.equal('/');
            should.exist(logEntry.req.headers);

            // response logging
            logEntry.res.statusCode.should.equal(500);
            should.exist(logEntry.res.responseTime);
            should.exist(logEntry.res.headers);

            // error logging
            logEntry.err.message.should.match(/Something\ blew\ up/);
            logEntry.err.name.should.equal('InternalServerError');
            logEntry.err.stack.should.match(/Something\ blew\ up/);

            done();
          });

        });

    });

    it('should display the application error message for non 500 errors', function (done) {

      app.use(koaJsonLogger());

      // 1st default test route that will catch uncaught downstream errors
      app.use(function *route1(next) {

        // throw a custom application error
        this.throw(400, 'Bad URL parameter format');
        yield next;
      });

      request(app.listen())
        .get('/')
        .expect(400)
        .end(function (err, res) {
          if (err) {
            should.not.exist(err);
            return done(err);
          }

          // should not leak out internal server error messages on 500
          // standard error response for the user
          res.text.should.equal('Bad URL parameter format');

          // read in log file entry
          fs.readFile('log/app_error.log', function (err, data) {
            if (err) {
              throw err;
            }

            // test JSON parsed log entry
            var logEntry = JSON.parse(data.toString());

            // bunyan property logging
            logEntry.name.should.equal('app');
            should.exist(logEntry.uid);
            logEntry.uid.should.match(/^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i);

            // request logging
            logEntry.req.method.should.equal('GET');
            logEntry.req.url.should.equal('/');
            should.exist(logEntry.req.headers);

            // response logging
            logEntry.res.statusCode.should.equal(400);
            should.exist(logEntry.res.responseTime);
            should.exist(logEntry.res.headers);

            // error logging
            logEntry.err.message.should.equal('Bad URL parameter format');
            logEntry.err.name.should.equal('BadRequestError');
            logEntry.err.stack.should.match(/Bad\ URL\ parameter\ format/);

            done();
          });

        });

    });

    it('should allow additional streams', function (done) {
      var stdErrStream = { level: 'error', stream: process.stderr };

      app.use(koaJsonLogger({ errStreams: [stdErrStream] }));

      // 1st default test route that will catch uncaught downstream errors
      app.use(function *route1(next) {

        // throw a custom application error
        this.throw(400, 'Bad URL parameter format');
        yield next;
      });

      request(app.listen())
        .get('/')
        .expect(400)
        .end(function (err, res) {
          if (err) {
            should.not.exist(err);
            return done(err);
          }

          // should not leak out internal server error messages on 500
          // standard error response for the user
          res.text.should.equal('Bad URL parameter format');

          // read in log file entry
          fs.readFile('log/app_error.log', function (err, data) {
            if (err) {
              throw err;
            }

            // test JSON parsed log entry
            var logEntry = JSON.parse(data.toString());

            // bunyan property logging
            logEntry.name.should.equal('app');
            should.exist(logEntry.uid);
            logEntry.uid.should.match(/^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i);

            // request logging
            logEntry.req.method.should.equal('GET');
            logEntry.req.url.should.equal('/');
            should.exist(logEntry.req.headers);

            // response logging
            logEntry.res.statusCode.should.equal(400);
            should.exist(logEntry.res.responseTime);
            should.exist(logEntry.res.headers);

            // error logging
            logEntry.err.message.should.equal('Bad URL parameter format');
            logEntry.err.name.should.equal('BadRequestError');
            logEntry.err.stack.should.match(/Bad\ URL\ parameter\ format/);

            done();
          });

        });

    });

    it('should default to 500 error status code', function (done) {

      app.use(koaJsonLogger());

      app.use(function *route1(next) {
        yield next;
        throw new Error('Oops! Something blew up.');
      });

      request(app.listen())
        .get('/')
        .expect(500)
        .end(function (err, res) {
          if (err) {
            should.not.exist(err);
            return done(err);
          }

          // should not leak out internal server error messages on 500
          // standard error response for the user
          res.text.should.equal('Internal Server Error');

          // read in log file entry
          fs.readFile('log/app_error.log', function (err, data) {
            if (err) {
              throw err;
            }

            // test JSON parsed log entry
            var logEntry = JSON.parse(data.toString());

            // bunyan property logging
            logEntry.name.should.equal('app');
            should.exist(logEntry.uid);
            logEntry.uid.should.match(/^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i);

            // request logging
            logEntry.req.method.should.equal('GET');
            logEntry.req.url.should.equal('/');
            should.exist(logEntry.req.headers);

            // response logging
            logEntry.res.statusCode.should.equal(500);
            should.exist(logEntry.res.responseTime);
            should.exist(logEntry.res.headers);

            // error logging
            logEntry.err.message.should.match(/Something\ blew\ up/);
            logEntry.err.name.should.equal('Error');
            logEntry.err.stack.should.match(/Something\ blew\ up/);

            done();
          });

        });

    });

    it('should respond with JSON/API 500 errors if response media type is JSON/API', function (done) {

      // Set the API response to JSON API format
      app.use(koaJsonLogger({
        jsonapi: true
      }));

      // 1st default test route that will catch uncaught downstream errors
      app.use(function *route1(next) {

        // throw a custom application error
        this.throw('Oops! Something blew up.');
        yield next;
      });

      request(app.listen())
        .get('/')
        .expect(500)
        .end(function (err, res) {
          if (err) {
            should.not.exist(err);
            return done(err);
          }

          // should not leak out internal server error messages on 500
          // standard error response for the user
          res.text.should.equal('{"status":500,"title":"Internal Server Error"}');

          // read in log file entry
          fs.readFile('log/app_error.log', function (err, data) {
            if (err) {
              throw err;
            }

            // test JSON parsed log entry
            var logEntry = JSON.parse(data.toString());

            // bunyan property logging
            logEntry.name.should.equal('app');
            should.exist(logEntry.uid);
            logEntry.uid.should.match(/^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i);

            // request logging
            logEntry.req.method.should.equal('GET');
            logEntry.req.url.should.equal('/');
            should.exist(logEntry.req.headers);

            // response logging
            logEntry.res.statusCode.should.equal(500);
            should.exist(logEntry.res.responseTime);
            logEntry.res.headers['content-type'].should.equal('application/vnd.api+json');

            // error logging
            logEntry.err.message.should.match(/Something\ blew\ up/);
            logEntry.err.name.should.equal('InternalServerError');
            logEntry.err.stack.should.match(/Something\ blew\ up/);

            done();
          });

        });

    });

    it('should emit conosle errors in development mode only', function (done) {

      process.env.NODE_ENV = 'development';

      app.emit = sinon.spy();

      app.use(koaJsonLogger());

      // 1st default test route that will catch uncaught downstream errors
      app.use(function *route1(next) {

        // throw a custom application error
        this.throw(400, 'Bad URL parameter format');
        yield next;
      });

      request(app.listen())
        .get('/')
        .expect(400)
        .end(function (err) {
          if (err) {
            should.not.exist(err);
            return done(err);
          }

          // test for console error output
          sinon.assert.calledOnce(app.emit);
          sinon.assert.calledWith(app.emit, 'error');

          done();
        });

    });

    it('should receive logs on stdout/stderr', function(done) {

      app.use(koaJsonLogger(
        {path: null}
      ));

      // default route for test
      app.use(function *route1(next) {
        this.body = 'Test Response is OK.';
        yield next;
      });

      request(app.listen())
        .get('/')
        .expect(200)
        .end(function (err, res) {
          if (err) {
            should.not.exist(err);
            return done(err);
          }

          // test http response
          res.text.should.equal('Test Response is OK.');


          fs.existsSync('log/app.log').should.be.false;

          done();

        });

    });

  });

});


lib/koa-json-logger.orig.js
'use strict';

var bunyan = require('bunyan'),
  util = require('util');

function reqSerializer(ctx) {
  return {
    url: ctx.url,
    headers: ctx.header,
    method: ctx.method,
    originalUrl: ctx.originalUrl,
    query: ctx.query
  };
}

function resSerializer(ctx) {

  var serializer = {
    statusCode: ctx.status,
    responseTime: ctx.responseTime
  };

  // if error, add the response body (if there is any)
  if (ctx.status >= 400 && ctx.body) {
    serializer.error = ctx.body;
  }

  return serializer;
}

module.exports = function koaLogger(opts) {

  opts = opts || {};

  var defaultName = opts.name || 'myapp';

  // default streams
  // log INFO and above to file
  // log ERROR and above to file
  // daily rotation
  // keep 3 back copies
  var streams = [
    {
      level: 'info',
      path: 'log/' + defaultName + '.log',
      period: '1d',
      count: 3
    }
  ];

  var defaultStreams = opts.streams || streams;

  var logger = bunyan.createLogger({

    name: defaultName,

    serializers: {
      req: reqSerializer,
      res: resSerializer,
      err: bunyan.stdSerializers.err
    },

    streams: defaultStreams
  });

  var logger2 = bunyan.createLogger({

    name: defaultName,

    serializers: {
      req: reqSerializer,
      res: resSerializer,
      err: bunyan.stdSerializers.err
    },

    streams: [
      {
        level: 'error',
        path: 'log/' + defaultName + '-error.log',
        period: '1d',
        count: 3
      }
    ]
  });

  return function *(next) {

    var start = new Date();

    var ctx = this;

    var done = function () {
      ctx.responseTime = new Date() - start;

      // log non error messages
      if (ctx.status < 400) {
        logger.info({req: ctx, res: ctx}, util.format('%s %s', ctx.method, ctx.originalUrl));
      }
      else {
        /*
        Application code can return an error without using "this.throw" or "ctx.throw"
        For example application code could do this and return:
          this.status = 403;
          this.body = 'Access denied';
        In this case the catch block below would not be triggered.
        We want this type of response to go to the error log
        Without this else block the req/res would silently not be logged, this else block should catch this edge case
        The disadvantage of not using "this.throw" in this manner is there is no error stack trace logged only the error message
        */
        logger2.error({req: ctx, res: ctx, err: ctx.body}, util.format('%s %s', ctx.method, ctx.originalUrl));
      }

    };

    ctx.res.once('finish', done);
    ctx.res.once('close', done);

    try {
      yield next;
    }
    catch (e) {

      console.log(e);

      // Response properties
      ctx.status = e.status;
      ctx.responseTime = new Date() - start;

      // log error message and stack trace
      logger2.error({req: ctx, res: ctx, err: e}, util.format('%s %s', ctx.method, ctx.originalUrl));

      // Do not leak internal server error message to the user
      // Friendly response message for user
      if (e.status === 500) {
        this.body = {status: 500, title: 'Internal Server Error'};
      }
      else {
        this.body = e.message;
      }

      // dump error to console in development mode
      //if (process.env.NODE_ENV === 'development') {
      //  if (e.status) {
      //    this.throw(e.status, e.message);
      //  }
      //  else {
      //    this.throw(500, 'Internal Server Error');
      //  }
      //}

    }

  };

};


lib/koa-json-logger.js
'use strict';

var path = require('path'),
  _ = require('lodash'),
  bunyan = require('bunyan'),
  uuid = require('uuid');

function reqSerializer(ctx) {
  return {
    url: ctx.url,
    headers: ctx.request.header,
    method: ctx.method,
    ip: ctx.ip,
    protocol: ctx.protocol,
    originalUrl: ctx.originalUrl,
    query: ctx.query
  };
}

function resSerializer(ctx) {
  return {
    statusCode: ctx.status,
    responseTime: ctx.responseTime,
    headers: ctx.response.header
  };
}

function uidSerializer (uuid) {
  return uuid;
}

var configDefaults = {
  name: 'app',
  path: 'log',
  jsonapi: false,
  isJsonApi: function () {
    return this.jsonapi;
  },
  outStreams: null,
  errStreams: null
};

module.exports = function koaLogger(opts) {

  opts = opts || {};

  var config = _.defaults(opts, configDefaults),
    env = process.env.NODE_ENV;

  var outStream = { level: 'info' };
  if (config.path === null) {
    outStream.stream = process.stdout;
  }
  else {
    outStream.path = path.join(config.path, config.name + '.log');
  }

  // Standard Logger
  var outLogger = bunyan.createLogger({

    name: config.name,

    serializers: _.extend({
      req: reqSerializer,
      res: resSerializer
    }, config.serializers),

    streams: [outStream].concat(config.outStreams || [])
  });

  var errStream = { level: 'error' };
  if (config.path === null) {
    errStream.stream = process.stderr;
  }
  else {
    errStream.path = path.join(config.path, config.name + '_error.log');
  }

  // Error Logger
  var errLogger = bunyan.createLogger({

    name: config.name,

    serializers: _.extend({
      uid: uidSerializer,
      req: reqSerializer,
      res: resSerializer,
      err: bunyan.stdSerializers.err
    }, config.serializers),

    streams: [errStream].concat(config.errStreams || [])
  });

  return function *logger(next) {

    var ctx = this,
      start = new Date();

    ctx.uuid = uuid.v4();

    // If logging for a JSON API set the response Content-type before logging is done so the header  is correctly logged
    if (config.isJsonApi()) {
      ctx.response.type = 'application/vnd.api+json';
    }

    try {
      yield next;

      ctx.responseTime = new Date() - start;

      outLogger.info({uid: ctx.uuid, req: ctx, res: ctx});
    }
    catch (err) {

      // Response properties
      ctx.status = err.status || 500;
      ctx.responseTime = new Date() - start;

      // log error message and stack trace
      errLogger.error({uid: ctx.uuid, req: ctx, res: ctx, err: err});

      // Handle 500 errors - do not leak internal server error message to the user.
      // Standard error response message for user
      if (ctx.status === 500) {
        (config.isJsonApi()) ? ctx.body = {status: 500, title: 'Internal Server Error'} : ctx.body = 'Internal Server Error';
      }
      else {
        ctx.body = err.details || err.message;
      }

      // Console output in development only
      if (env === 'development') {
        ctx.app.emit('error', err, this);
      }

    }

    /*
    Currently - if a nested object is thrown the Content-type is set to 'application/json'
    Example:
    this.throw(401, {
      message: {
        status: 401,
        title: 'Unauthorized',
        detail: 'No Authorization header found'
      }
    })
    If JSON API is enabled to work around this we need to set the response Content-type again
    */
    if (config.isJsonApi()) {
      ctx.response.type = 'application/vnd.api+json';
    }

  };

};

module.exports.serializers = {
  uid: uidSerializer,
  req: reqSerializer,
  res: resSerializer,
  err: bunyan.stdSerializers.err
};
